anybox/recipe/odoo/base.py

```diff
@@ -279,14 +301,144 @@ def install_recipe_requirements(self):
+
+    def list_develops(self):
+        """At any point in time, list the projects that have been developed.
+
+        This can work as soon as the recipe is instantiated
+        (because the 'buildout' part) has already been executed.
+        In particular, it does not rely on the workingset init done by
+        :class:`zc.buildout.easy_install.Installer` and
+        can be used in precedence rules that need to be executed before calling
+        the Installer indirectly via ``zc.recipe.eggs``
+
+        :return: list of project names
+
+        Implementation simply lists the develop eggs directory
+        There's probably better to be done.
+        """
+        devdir_contents = (f.rsplit('.', 1) for f in os.listdir(
+            self.b_options['develop-eggs-directory']))
+        return [s[0] for s in devdir_contents if s[1] == 'egg-link']
+
+    def apply_odoo_requirements_file(self):
+        """Try and read Odoo's 'requirements.txt' and apply it.
+
+        This file appeared in the course of Odoo 8 lifetime. If not available,
+        a warning is issued, that's all.
+
+        Entries from the requirements file are applied if there is not already
+        an entry in the versions section for the same project.
+
+        A more interesting behaviour would be to apply then if they don't
+        contradict an existing entry in the versions section, but that's far
+        more complicated.
+        """
+        req_fname = 'requirements.txt'
+        req_path = join(self.openerp_dir, req_fname)
+        if not os.path.exists(req_path):
+            logger.warn("%r not found in this version of "
+                        "Odoo, although the configuration said to use it. "
+                        "Proceeding anyway.", req_fname)
+        else:
+            # pip wouldn't be importable before the call to
+            # install_recipe_requirements()
+
+            # if an extension has used pip before, it can be left in a
+            # strange state where pip.req is not usable nor reloadable
+            # anymore. (may have something to do with the fact that the
+            # first import is done from a tmp dir
+            # that does not exist any more).
+            # So, better to clean that before hand.
+            for k in sys.modules.keys():
+                if k.split('.', 1)[0] == 'pip':
+                    del sys.modules[k]
+
+            # it is useless to mutate the versions section at this point
+            # it's already been used to populate the Installer class variable
+            versions = Installer._versions
+            develops = self.list_develops()
+
+            new_reqs = set()
+            from pip.req import parse_requirements
+            # pip internals are protected against the fact of not passing
+            # a session with ``is None``. OTOH, the session is not used
+            # if the file is local (direct path, not an URL), so we cheat
+            # it.
+            fake_session = object()
```

Comment:
This is fine as long as there are no "-r/--requirement <URL>" line in the requirements.txt file.